### PR TITLE
rbd-mirror: workaround for intermingled lockdep singletons

### PR DESCRIPTION
**Pull request by dillaman**
If librados constructs the CephContext, lockdep will be initialized
in both the librados shared address space and also within the app
address space.  This breaks lockdep's functionality and results in
a crash upon shutdown.

Signed-off-by: Jason Dillaman <dillaman@redhat.com>

### CODE REVIEW

**Comment by dillaman** [comment]
@trociny @rjfd FYI -- after some digging, I created this workaround for the lockdep issues.  Short term I think this will resolve the crashes on shutdown when lockdep is enabled.

**Comment by rjfd** [comment]
@dillaman lgtm --- already tested and the assertion failure is no more.
I was wondering if the only option to make lockdep truly singleton would be to make it a library on its own.

**Comment by dillaman** [comment]
@rjfd Yeah -- we run into lots of issues due to dual linkage of libcommon/libosdc/etc in librados and librbd/other apps.

**Comment by jdurgin** [comment]
This reminds me we might want to use CINIT_FLAG_UNPRIVILEGED_DAEMON_DEFAULTS for common_preinit() to get logs and asok by default for the remote cluster too.

**Comment by trociny** [comment]
@dillaman Good idea, but there is an issue with the patch: by some reason command line options are not applied. You can see this running rbd_mirror.sh with NOCLEANUP, and observe that asok and log files are not created in tmp directory but in run/cluster2/out directory (as specified in cluster2.conf).

So far I have failed to understand why.

File: src/tools/rbd_mirror/Replayer.cc [hunk 1]
```diff
@@ -143,37 +146,59 @@ int Replayer::init()
...
+    derr << "could not parse environment for " << m_peer << ":"
+         << cpp_strerror(r) << dendl;
+    cct->put();
     return r;
   }
 
   if (!m_args.empty()) {
-    r = m_remote->conf_parse_argv(m_args.size(), &m_args[0]);
+    // librados::Rados::conf_parse_argv
+    r = cct->_conf->parse_argv(args);
```
Comment (trociny) [inline src/tools/rbd_mirror/Replayer.cc]: @dillaman s/args/m_args/ :-)

**Comment by dillaman** [comment]
Updated and fixed -- thx.

**Comment by trociny** [comment]
lgtm

**Pull request merged**